### PR TITLE
Update integration part with NNCF

### PR DESCRIPTION
**Pull request by kshpv**

### CODE REVIEW

**Comment by druzhkov-paul** [comment]
run tests

File: mmdet/core/utils/misc.py [hunk 1]
```diff
@@ -93,20 +95,22 @@ def topk(x, k, dim=None, **kwargs):
...
 
-    if is_in_onnx_export():
+    if is_in_onnx_export() or is_in_nncf_tracing():
         n = operators.shape_as_tensor(x)[dim].unsqueeze(0)
         if not isinstance(k, torch.Tensor):
             k = torch.tensor([k], dtype=torch.long)
         # Workaround for ONNXRuntime: convert values to int to get minimum.
-        n = torch.min(torch.cat((k, n), dim=0).int()).long()
+        with no_nncf_trace():
+            n = torch.min(torch.cat((k, n), dim=0).int()).long()
```
Comment (druzhkov-paul) [inline mmdet/core/utils/misc.py:104]: Please put a comment before the line it relates to, not the context manager.

File: mmdet/integration/nncf/compression.py [hunk 2]
```diff
@@ -207,59 +210,14 @@ def wrap_inputs(args, kwargs):
...
-        return model
-    module_.op.run_export_quantization = partial(run_hacked_export_quantization, module_.op)
-    logger.info('Change an export function for the first Conv to avoid saturation issue on AVX2, AVX512')
-    return model
-
-
 def get_uncompressed_model(module):
     if not is_nncf_enabled():
         return module
-    from nncf.nncf_network import NNCFNetwork
```
Comment (druzhkov-paul) [inline mmdet/integration/nncf/compression.py]: Isn't it required?

File: mmdet/models/detectors/single_stage.py [hunk 3]
```diff
@@ -104,14 +104,15 @@ def simple_test(self, img, img_metas, rescale=False, postprocess=True):
...
-        if torch.onnx.is_in_onnx_export():
+        if torch.onnx.is_in_onnx_export() or is_in_nncf_tracing():
             return bbox_results[0]
 
         if postprocess:
-            bbox_results = [
-                self.postprocess(det_bboxes, det_labels, None, img_metas, rescale=rescale)
-                for det_bboxes, det_labels in bbox_results
-            ]
+            with no_nncf_trace():
```
Comment (druzhkov-paul) [inline mmdet/models/detectors/single_stage.py]: Seems that this code is unreachable in NNCF tracing mode at all (see the return statement a couple of lines above).

File: mmdet/models/detectors/two_stage.py [hunk 4]
```diff
@@ -190,14 +190,11 @@ def simple_test(self,
...
         """Test without augmentation."""
         assert self.with_bbox, 'Bbox head must be implemented.'
-
         x = self.extract_feat(img)
-
         if proposals is None:
             proposal_list = self.rpn_head.simple_test_rpn(x, img_metas)
         else:
             proposal_list = proposals
-
```
Comment (druzhkov-paul) [inline mmdet/models/detectors/two_stage.py]: Don't like empty spaces? :smile:

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 5]
```diff
@@ -225,46 +227,47 @@ def loss(self,
...
-                if self.reg_decoded_bbox:
-                    # When the regression loss (e.g. `IouLoss`,
-                    # `GIouLoss`, `DIouLoss`) is applied directly on
-                    # the decoded bounding boxes, it decodes the
-                    # already encoded coordinates to absolute format.
-                    bbox_pred = self.bbox_coder.decode(rois[:, 1:], bbox_pred)
-                if self.reg_class_agnostic:
-                    pos_bbox_pred = bbox_pred.view(
-                        bbox_pred.size(0), 4)[pos_inds.type(torch.bool)]
+        with no_nncf_trace():
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:229]: Just out of curiosity, is there available a decorator that disables NNCF tracing inside a whole function's body?

File: mmdet/models/roi_heads/standard_roi_head_with_text.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 import torch
 
 from mmdet.core import bbox2roi, build_assigner, build_sampler
+from mmdet.integration.nncf.utils import is_in_nncf_tracing
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/standard_roi_head_with_text.py:7]: Is it used?

File: mmdet/models/roi_heads/test_mixins.py [hunk 7]
```diff
@@ -7,6 +7,8 @@
                         merge_aug_masks, multiclass_nms)
 
 from mmdet.core.utils.misc import dummy_pad
+from mmdet.integration.nncf.utils import is_in_nncf_tracing
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/test_mixins.py]:
```suggestion
```

File: mmdet/models/roi_heads/test_mixins.py [hunk 8]
```diff
@@ -58,19 +60,22 @@ def simple_test_bboxes(self,
                            rcnn_test_cfg,
                            rescale=False):
         """Test only det bboxes without augmentation."""
-        rois = bbox2roi(proposals)
+        from mmdet.integration.nncf.utils import no_nncf_trace
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/test_mixins.py]:
```suggestion
```

**Review by druzhkov-paul — COMMENTED** [review]
And please remove `mmdet/models/roi_heads/roi_extractors/single_level.py`.

File: mmdet/core/utils/misc.py (hunk 1 above)
Comment (kshpv) [reply mmdet/core/utils/misc.py:104]: Done

File: mmdet/integration/nncf/compression.py (hunk 2 above)
Comment (kshpv) [reply mmdet/integration/nncf/compression.py]: yes, as it is used in the following `isisntance() `function

File: mmdet/models/detectors/single_stage.py (hunk 3 above)
Comment (kshpv) [reply mmdet/models/detectors/single_stage.py]: I think so but remained this context to be sure that postprocessing will be non-traced. If you also think that is unreachable, I removed this

File: mmdet/models/detectors/two_stage.py (hunk 4 above)
Comment (kshpv) [reply mmdet/models/detectors/two_stage.py]: Just random changes. I reverted them back

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py (hunk 5 above)
Comment (kshpv) [reply mmdet/models/roi_heads/bbox_heads/bbox_head.py:229]: Currently, no

File: mmdet/models/roi_heads/test_mixins.py (hunk 7 above)
Comment (kshpv) [reply mmdet/models/roi_heads/test_mixins.py]: Done

File: mmdet/models/roi_heads/test_mixins.py (hunk 8 above)
Comment (kshpv) [reply mmdet/models/roi_heads/test_mixins.py]: Done

File: mmdet/models/roi_heads/standard_roi_head_with_text.py (hunk 6 above)
Comment (kshpv) [reply mmdet/models/roi_heads/standard_roi_head_with_text.py:7]: Yes, in lines 169, 197

**Comment by kshpv** [comment]
run tests

File: mmdet/integration/nncf/compression.py [hunk 9]
```diff
@@ -216,6 +216,7 @@ def wrap_inputs(args, kwargs):
 
 
 def get_uncompressed_model(module):
+    from nncf.nncf_network import NNCFNetwork
     if not is_nncf_enabled():
         return module
```
Comment (druzhkov-paul) [inline mmdet/integration/nncf/compression.py]:
```suggestion
    if not is_nncf_enabled():
        return module
    from nncf.nncf_network import NNCFNetwork
```

File: mmdet/models/dense_heads/fcos_head.py [hunk 10]
```diff
@@ -5,9 +5,12 @@
 from mmcv.runner import force_fp32
 
 from mmdet.core import distance2bbox, multi_apply, multiclass_nms, reduce_mean
+from mmdet.integration.nncf.utils import is_in_nncf_tracing
 from ..builder import HEADS, build_loss
 from .anchor_free_head import AnchorFreeHead
 
+
+
```
Comment (druzhkov-paul) [inline mmdet/models/dense_heads/fcos_head.py]:
```suggestion
```

File: mmdet/models/dense_heads/yolo_head.py [hunk 11]
```diff
@@ -11,11 +11,13 @@
 from mmdet.core import (build_anchor_generator, build_assigner,
                         build_bbox_coder, build_sampler, images_to_levels,
                         multi_apply, multiclass_nms)
+from mmdet.integration.nncf.utils import is_in_nncf_tracing
 from ..builder import HEADS, build_loss
 from .base_dense_head import BaseDenseHead
 from .dense_test_mixins import BBoxTestMixin
 
 
+
```
Comment (druzhkov-paul) [inline mmdet/models/dense_heads/yolo_head.py]:
```suggestion
```

File: mmdet/models/roi_heads/htc_roi_head.py [hunk 12]
```diff
@@ -5,10 +5,13 @@
                         merge_aug_masks, multiclass_nms)
 from mmdet.core.mask.transforms import mask2result
 from mmdet.core.utils.misc import dummy_pad
+from mmdet.integration.nncf.utils import is_in_nncf_tracing
 from ..builder import HEADS, build_head, build_roi_extractor
 from .cascade_roi_head import CascadeRoIHead
 
 
+
+
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/htc_roi_head.py]:
```suggestion
```

File: mmdet/models/roi_heads/scnet_roi_head.py [hunk 13]
```diff
@@ -5,10 +5,12 @@
                         merge_aug_masks, multiclass_nms)
 from mmdet.core.mask.transforms import mask2result
 from mmdet.core.utils.misc import dummy_pad
+from mmdet.integration.nncf.utils import is_in_nncf_tracing
 from ..builder import HEADS, build_head, build_roi_extractor
 from .cascade_roi_head import CascadeRoIHead
 
 
+
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/scnet_roi_head.py]:
```suggestion
```

File: mmdet/models/roi_heads/standard_roi_head_with_text.py [hunk 14]
```diff
@@ -10,6 +10,7 @@
 from .test_mixins import dummy_pad
 
 
+
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/standard_roi_head_with_text.py]:
```suggestion
```

File: mmdet/integration/nncf/__init__.py [hunk 15]
```diff
@@ -1,7 +1,7 @@
 from .compression import (check_nncf_is_enabled, get_nncf_config_from_meta,
                           get_nncf_metadata, get_uncompressed_model,
                           is_checkpoint_nncf, wrap_nncf_model)
-from .compression_hooks import CompressionHook
+from .compression_hooks import CompressionHook, CheckpointHookBeforeTraining
```
Comment (druzhkov-paul) [inline mmdet/integration/nncf/__init__.py:4]: Please add `CheckpointHookBeforeTraining` to `__all__`.

**Comment by druzhkov-paul** [comment]
run tests

**Comment by druzhkov-paul** [comment]
run tests

File: mmdet/integration/nncf/__init__.py (hunk 15 above)
Comment (kshpv) [reply mmdet/integration/nncf/__init__.py:4]: Done

File: mmdet/models/roi_heads/standard_roi_head_with_text.py (hunk 14 above)
Comment (kshpv) [reply mmdet/models/roi_heads/standard_roi_head_with_text.py]: Done

File: mmdet/models/roi_heads/scnet_roi_head.py (hunk 13 above)
Comment (kshpv) [reply mmdet/models/roi_heads/scnet_roi_head.py]: Done

File: mmdet/models/roi_heads/htc_roi_head.py (hunk 12 above)
Comment (kshpv) [reply mmdet/models/roi_heads/htc_roi_head.py]: Done

File: mmdet/models/dense_heads/yolo_head.py (hunk 11 above)
Comment (aleksei-kashapov) [reply mmdet/models/dense_heads/yolo_head.py]: Done

File: mmdet/models/dense_heads/fcos_head.py (hunk 10 above)
Comment (kshpv) [reply mmdet/models/dense_heads/fcos_head.py]: Done

File: mmdet/integration/nncf/compression.py (hunk 9 above)
Comment (kshpv) [reply mmdet/integration/nncf/compression.py]: Done

**Comment by kshpv** [comment]
run tests

**Review by druzhkov-paul — APPROVED** [review]

**Comment by druzhkov-paul** [comment]
Merging this as long as all tests here and in OTE (https://github.com/openvinotoolkit/training_extensions/pull/533) passed.

**Pull request merged**